// web-app/src/components/AuthModule.jsx
import React, { useState } from 'react';
import { auth } from '../../../Backend/config/firebase';
import { 
  signInWithEmailAndPassword, 
  createUserWithEmailAndPassword, 
  sendPasswordResetEmail 
} from 'firebase/auth';
import { userService } from '../../../Backend/services/userService';
import { useAppStore } from '../../../Backend/store/useAppStore';
import { KeyRound, Mail, User, Phone, ShieldCheck, ArrowLeft, Loader2, Sparkles } from 'lucide-react';
import './AuthModule.css';

export default function AuthModule() {
  const setUser = useAppStore((state) => state.setUser);
  
  const [view, setView] = useState('landing'); // 'landing' | 'login' | 'signup' | 'forgot' | 'verify'
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const [name, setName] = useState('');
  const [phone, setPhone] = useState('');
  const [otp, setOtp] = useState('');
  
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');
  const [simulatedOtp, setSimulatedOtp] = useState('');
  const [tempUser, setTempUser] = useState(null); // Save user object during verification step

  const handleLogin = async (e) => {
    e.preventDefault();
    if (!email || !password) {
      setError('Please fill in all fields.');
      return;
    }
    setLoading(true);
    setError('');
    try {
      const userCredential = await signInWithEmailAndPassword(auth, email, password);
      const profile = await userService.getUserProfile(userCredential.user.uid);
      
      if (profile && profile.isProfileVerified) {
        setUser({ uid: userCredential.user.uid, email, ...profile });
      } else {
        // Needs profile verification
        setTempUser(userCredential.user);
        setView('verify');
      }
    } catch (err) {
      console.error(err);
      setError(err.message.replace('Firebase:', '').trim());
    } finally {
      setLoading(false);
    }
  };

  const handleSignup = async (e) => {
    e.preventDefault();
    if (!email || !password || !confirmPassword) {
      setError('Please fill in all fields.');
      return;
    }
    if (password !== confirmPassword) {
      setError('Passwords do not match.');
      return;
    }
    if (password.length < 6) {
      setError('Password must be at least 6 characters.');
      return;
    }
    setLoading(true);
    setError('');
    try {
      const userCredential = await createUserWithEmailAndPassword(auth, email, password);
      setTempUser(userCredential.user);
      setView('verify');
    } catch (err) {
      console.error(err);
      setError(err.message.replace('Firebase:', '').trim());
    } finally {
      setLoading(false);
    }
  };

  const handleForgotPassword = async (e) => {
    e.preventDefault();
    if (!email) {
      setError('Please enter your email address.');
      return;
    }
    setLoading(true);
    setError('');
    try {
      await sendPasswordResetEmail(auth, email);
      alert('Password reset link sent to your email.');
      setView('login');
    } catch (err) {
      console.error(err);
      setError(err.message.replace('Firebase:', '').trim());
    } finally {
      setLoading(false);
    }
  };

  const handleSendOtp = () => {
    if (!name.trim()) {
      setError('Please enter your Full Name.');
      return;
    }
    if (!phone.trim() || phone.length < 10) {
      setError('Please enter a valid Phone Number.');
      return;
    }

    setLoading(true);
    setError('');
    
    // Simulate sending SMS
    setTimeout(() => {
      const generatedOtp = Math.floor(1000 + Math.random() * 9000).toString();
      setSimulatedOtp(generatedOtp);
      setLoading(false);
      
      // Visual Alert for simulated OTP
      alert(`[SMS SIMULATION] Your TrackBack verification code is: ${generatedOtp}`);
    }, 1200);
  };

  const handleVerifyOtp = async (e) => {
    e.preventDefault();
    if (!otp) {
      setError('Please enter the 4-digit code.');
      return;
    }
    if (otp !== simulatedOtp) {
      setError('Invalid verification code. Please try again.');
      return;
    }

    setLoading(true);
    setError('');
    try {
      const uid = tempUser.uid;
      const profileData = {
        name: name.trim(),
        phone: phone.trim(),
        phoneNumber: phone.trim(),
        isProfileVerified: true,
        photoURL: `https://api.dicebear.com/7.x/adventurer/svg?seed=${encodeURIComponent(name)}`
      };

      await userService.updateUserProfile(uid, profileData);
      
      setUser({
        uid,
        email: tempUser.email,
        ...profileData
      });
    } catch (err) {
      console.error(err);
      setError('Verification failed: ' + err.message);
    } finally {
      setLoading(false);
    }
  };

  const renderLanding = () => (
    <div className="auth-card glass" style={{ textAlign: 'center' }}>
      <div className="auth-logo">
        <Sparkles size={36} />
        <span>TrackBack</span>
      </div>
      <p style={{ color: 'var(--light-text)', margin: '1rem 0 2.5rem', lineHeight: '1.6' }}>
        A secure, AI-powered lost and found platform to retrieve your valuables safely and prevent fraud.
      </p>
      
      <div style={{ display: 'flex', flexDirection: 'column', gap: '1rem' }}>
        <button className="btn btn-primary" onClick={() => setView('login')}>
          Sign In
        </button>
        <button className="btn btn-secondary" onClick={() => setView('signup')}>
          Create Account
        </button>
      </div>
    </div>
  );

  const renderLogin = () => (
    <div className="auth-card glass">
      <button className="btn btn-secondary" style={{ padding: '0.5rem', borderRadius: '50%', marginBottom: '1.5rem' }} onClick={() => setView('landing')}>
        <ArrowLeft size={18} />
      </button>
      
      <div className="auth-header">
        <h2 className="auth-title">Welcome Back</h2>
        <p className="auth-subtitle">Sign in to search, report or claim items</p>
      </div>

      {error && <div className="auth-error">{error}</div>}

      <form onSubmit={handleLogin}>
        <div className="form-group">
          <label className="form-label">Email Address</label>
          <div style={{ position: 'relative' }}>
            <Mail size={16} style={{ position: 'absolute', left: '12px', top: '15px', color: 'var(--light-text)' }} />
            <input 
              type="email" 
              className="form-input" 
              style={{ paddingLeft: '2.5rem' }}
              placeholder="name@example.com" 
              value={email}
              onChange={(e) => setEmail(e.target.value)}
              required
            />
          </div>
        </div>

        <div className="form-group">
          <label className="form-label">Password</label>
          <div style={{ position: 'relative' }}>
            <KeyRound size={16} style={{ position: 'absolute', left: '12px', top: '15px', color: 'var(--light-text)' }} />
            <input 
              type="password" 
              className="form-input" 
              style={{ paddingLeft: '2.5rem' }}
              placeholder="••••••••" 
              value={password}
              onChange={(e) => setPassword(e.target.value)}
              required
            />
          </div>
        </div>

        <button 
          className="auth-link forgot-password-link" 
          type="button"
          onClick={() => setView('forgot')}
        >
          Forgot Password?
        </button>

        <button className="btn btn-primary" style={{ width: '100%', marginTop: '1rem' }} type="submit" disabled={loading}>
          {loading ? <Loader2 className="spinner" size={18} style={{ borderLeftColor: '#fff', borderTopColor: 'transparent', width: 18, height: 18 }} /> : 'Sign In'}
        </button>
      </form>

      <div className="auth-footer">
        Don't have an account?{' '}
        <button className="auth-link" onClick={() => setView('signup')}>Sign Up</button>
      </div>
    </div>
  );

  const renderSignup = () => (
    <div className="auth-card glass">
      <button className="btn btn-secondary" style={{ padding: '0.5rem', borderRadius: '50%', marginBottom: '1.5rem' }} onClick={() => setView('landing')}>
        <ArrowLeft size={18} />
      </button>

      <div className="auth-header">
        <h2 className="auth-title">Create Account</h2>
        <p className="auth-subtitle">Join the community to find and recover items</p>
      </div>

      {error && <div className="auth-error">{error}</div>}

      <form onSubmit={handleSignup}>
        <div className="form-group">
          <label className="form-label">Email Address</label>
          <div style={{ position: 'relative' }}>
            <Mail size={16} style={{ position: 'absolute', left: '12px', top: '15px', color: 'var(--light-text)' }} />
            <input 
              type="email" 
              className="form-input" 
              style={{ paddingLeft: '2.5rem' }}
              placeholder="name@example.com" 
              value={email}
              onChange={(e) => setEmail(e.target.value)}
              required
            />
          </div>
        </div>

        <div className="form-group">
          <label className="form-label">Password</label>
          <div style={{ position: 'relative' }}>
            <KeyRound size={16} style={{ position: 'absolute', left: '12px', top: '15px', color: 'var(--light-text)' }} />
            <input 
              type="password" 
              className="form-input" 
              style={{ paddingLeft: '2.5rem' }}
              placeholder="•••••••• (Min 6 chars)" 
              value={password}
              onChange={(e) => setPassword(e.target.value)}
              required
            />
          </div>
        </div>

        <div className="form-group">
          <label className="form-label">Confirm Password</label>
          <div style={{ position: 'relative' }}>
            <KeyRound size={16} style={{ position: 'absolute', left: '12px', top: '15px', color: 'var(--light-text)' }} />
            <input 
              type="password" 
              className="form-input" 
              style={{ paddingLeft: '2.5rem' }}
              placeholder="••••••••" 
              value={confirmPassword}
              onChange={(e) => setConfirmPassword(e.target.value)}
              required
            />
          </div>
        </div>

        <button className="btn btn-primary" style={{ width: '100%', marginTop: '1.5rem' }} type="submit" disabled={loading}>
          {loading ? <Loader2 className="spinner" size={18} style={{ borderLeftColor: '#fff', borderTopColor: 'transparent', width: 18, height: 18 }} /> : 'Sign Up'}
        </button>
      </form>

      <div className="auth-footer">
        Already have an account?{' '}
        <button className="auth-link" onClick={() => setView('login')}>Sign In</button>
      </div>
    </div>
  );

  const renderForgot = () => (
    <div className="auth-card glass">
      <button className="btn btn-secondary" style={{ padding: '0.5rem', borderRadius: '50%', marginBottom: '1.5rem' }} onClick={() => setView('login')}>
        <ArrowLeft size={18} />
      </button>

      <div className="auth-header">
        <h2 className="auth-title">Reset Password</h2>
        <p className="auth-subtitle">We will send you a password reset link</p>
      </div>

      {error && <div className="auth-error">{error}</div>}

      <form onSubmit={handleForgotPassword}>
        <div className="form-group">
          <label className="form-label">Email Address</label>
          <div style={{ position: 'relative' }}>
            <Mail size={16} style={{ position: 'absolute', left: '12px', top: '15px', color: 'var(--light-text)' }} />
            <input 
              type="email" 
              className="form-input" 
              style={{ paddingLeft: '2.5rem' }}
              placeholder="name@example.com" 
              value={email}
              onChange={(e) => setEmail(e.target.value)}
              required
            />
          </div>
        </div>

        <button className="btn btn-primary" style={{ width: '100%', marginTop: '1.5rem' }} type="submit" disabled={loading}>
          {loading ? <Loader2 className="spinner" size={18} style={{ borderLeftColor: '#fff', borderTopColor: 'transparent', width: 18, height: 18 }} /> : 'Send Reset Link'}
        </button>
      </form>
    </div>
  );

  const renderVerify = () => (
    <div className="auth-card glass">
      <div className="auth-header">
        <ShieldCheck size={44} style={{ color: 'var(--primary-color)', marginBottom: '0.75rem' }} />
        <h2 className="auth-title">Verify Profile</h2>
        <p className="auth-subtitle">All users must complete profile setup to interact on the platform.</p>
      </div>

      {error && <div className="auth-error">{error}</div>}

      {!simulatedOtp ? (
        <div>
          <div className="form-group">
            <label className="form-label">Full Name</label>
            <div style={{ position: 'relative' }}>
              <User size={16} style={{ position: 'absolute', left: '12px', top: '15px', color: 'var(--light-text)' }} />
              <input 
                type="text" 
                className="form-input" 
                style={{ paddingLeft: '2.5rem' }}
                placeholder="John Doe" 
                value={name}
                onChange={(e) => setName(e.target.value)}
                required
              />
            </div>
          </div>

          <div className="form-group">
            <label className="form-label">Mobile Number</label>
            <div style={{ position: 'relative', display: 'flex', gap: '0.5rem' }}>
              <div className="form-input" style={{ width: '80px', display: 'flex', alignItems: 'center', justifyContent: 'center', fontWeight: 'bold', background: '#eaeaea' }}>
                +91
              </div>
              <div style={{ position: 'relative', flexGrow: 1 }}>
                <Phone size={16} style={{ position: 'absolute', left: '12px', top: '15px', color: 'var(--light-text)' }} />
                <input 
                  type="tel" 
                  className="form-input" 
                  style={{ paddingLeft: '2.5rem' }}
                  placeholder="00000 00000" 
                  value={phone}
                  onChange={(e) => setPhone(e.target.value)}
                  maxLength={10}
                  required
                />
              </div>
            </div>
          </div>

          <button className="btn btn-primary" style={{ width: '100%', marginTop: '1.5rem' }} onClick={handleSendOtp} disabled={loading}>
            {loading ? <Loader2 className="spinner" size={18} style={{ borderLeftColor: '#fff', borderTopColor: 'transparent', width: 18, height: 18 }} /> : 'Send Verification OTP'}
          </button>
        </div>
      ) : (
        <form onSubmit={handleVerifyOtp}>
          <div className="form-group" style={{ textAlign: 'center' }}>
            <label className="form-label">Enter 4-Digit Code</label>
            <p style={{ fontSize: '0.8rem', color: 'var(--primary-color)', marginBottom: '0.5rem' }}>Sent to +91 {phone}</p>
            <input 
              type="text" 
              className="form-input" 
              style={{ fontSize: '2rem', letterSpacing: '8px', textAlign: 'center', fontWeight: 'bold' }}
              placeholder="0000" 
              maxLength={4}
              value={otp}
              onChange={(e) => setOtp(e.target.value)}
              required
            />
          </div>

          <button className="btn btn-primary" style={{ width: '100%', marginTop: '1.5rem' }} type="submit" disabled={loading}>
            {loading ? <Loader2 className="spinner" size={18} style={{ borderLeftColor: '#fff', borderTopColor: 'transparent', width: 18, height: 18 }} /> : 'Verify & Continue'}
          </button>

          <button 
            className="auth-link" 
            style={{ display: 'block', margin: '1.5rem auto 0', fontSize: '0.9rem' }} 
            type="button" 
            onClick={() => { setSimulatedOtp(''); setOtp(''); }}
          >
            Change Phone Number
          </button>
        </form>
      )}
    </div>
  );

  return (
    <div className="auth-wrapper">
      <div className="auth-bg-circle circle-1"></div>
      <div className="auth-bg-circle circle-2"></div>
      {view === 'landing' && renderLanding()}
      {view === 'login' && renderLogin()}
      {view === 'signup' && renderSignup()}
      {view === 'forgot' && renderForgot()}
      {view === 'verify' && renderVerify()}
    </div>
  );
}
