// web-app/src/components/ReportWizard.jsx
import React, { useState, useRef } from 'react';
import { useAppStore } from '../../../Backend/store/useAppStore';
import { aiService } from '../../../Backend/services/aiService';
import { itemService } from '../../../Backend/services/itemService';
import { Upload, X, Loader2, ShieldCheck, AlertCircle, Sparkles, HelpCircle, Check, ArrowRight } from 'lucide-react';
import './Wizard.css';

export default function ReportWizard({ onClose }) {
  const user = useAppStore((state) => state.user);
  
  const [step, setStep] = useState(1);
  const [type, setType] = useState('found'); // 'found' | 'lost'
  const [imageFile, setImageFile] = useState(null);
  const [imagePreview, setImagePreview] = useState('');
  
  // Form fields
  const [title, setTitle] = useState('');
  const [category, setCategory] = useState('Electronics');
  const [description, setDescription] = useState('');
  const [location, setLocation] = useState('');
  const [phone, setPhone] = useState(user?.phone || '');
  
  // Verification questions (Found only)
  const [questions, setQuestions] = useState([
    { q: '', a: '' },
    { q: '', a: '' },
    { q: '', a: '' }
  ]);

  // Status & loading flags
  const [moderationLoading, setModerationLoading] = useState(false);
  const [moderationResult, setModerationResult] = useState(null); // { verified, reason }
  const [questionsLoading, setQuestionsLoading] = useState(false);
  const [publishLoading, setPublishLoading] = useState(false);
  const [error, setError] = useState('');

  const fileInputRef = useRef(null);
  const categories = ['Electronics', 'Wallets & Purses', 'Keys', 'Documents', 'Pets', 'Other'];

  const handleImageChange = async (e) => {
    const file = e.target.files[0];
    if (!file) return;

    setImageFile(file);
    setImagePreview(URL.createObjectURL(file));
    setError('');

    // If item type is Found, perform Image Moderation right away
    if (type === 'found') {
      setModerationLoading(true);
      setModerationResult(null);
      try {
        const result = await aiService.moderateImage(file);
        setModerationResult(result);
        if (!result.verified) {
          setError(result.reason || 'Image is blocked by moderation rules.');
        } else {
          // Trigger question generation in parallel or proceed
          generateAIQuestions(file);
        }
      } catch (err) {
        console.error(err);
        setError('Image moderation failed. Please try again.');
      } finally {
        setModerationLoading(false);
      }
    }
  };

  const generateAIQuestions = async (file) => {
    setQuestionsLoading(true);
    try {
      const generated = await aiService.generateQuestions(file);
      if (generated && generated.length === 3) {
        setQuestions(generated);
      }
    } catch (err) {
      console.error('Failed to generate questions, using fallbacks', err);
    } finally {
      setQuestionsLoading(false);
    }
  };

  const handleRemoveImage = () => {
    setImageFile(null);
    setImagePreview('');
    setModerationResult(null);
    setError('');
    setQuestions([
      { q: '', a: '' },
      { q: '', a: '' },
      { q: '', a: '' }
    ]);
  };

  const handleNext = () => {
    if (step === 1) {
      if (type === 'found' && !imageFile) {
        setError('Please upload an image of the found item.');
        return;
      }
      if (type === 'found' && moderationResult && !moderationResult.verified) {
        setError('Please upload a valid item image. The current image was rejected.');
        return;
      }
      setStep(2);
    } else if (step === 2) {
      if (!title.trim() || !description.trim()) {
        setError('Please fill in both title and description.');
        return;
      }
      setStep(3);
    } else if (step === 3) {
      if (!location.trim() || !phone.trim()) {
        setError('Please fill in both location and phone number.');
        return;
      }
      setError('');
      if (type === 'found') {
        setStep(4);
      } else {
        setStep(5); // Skip questions for lost items
      }
    } else if (step === 4) {
      // Validate questions
      const emptyQuestion = questions.some(q => !q.q.trim() || !q.a.trim());
      if (emptyQuestion) {
        setError('Please complete all 3 verification questions and answers.');
        return;
      }
      setError('');
      setStep(5);
    }
  };

  const handleBack = () => {
    setError('');
    if (step === 5 && type === 'lost') {
      setStep(3);
    } else {
      setStep(step - 1);
    }
  };

  const handlePublish = async () => {
    setPublishLoading(true);
    setError('');
    
    try {
      const itemData = {
        title: title.trim(),
        type,
        category,
        description: description.trim(),
        location: location.trim(),
        phoneNumber: phone.trim(),
        userId: user.uid,
        user: user.name,
        image: imageFile, // This will be handled inside itemService.addItem
      };

      if (type === 'found') {
        itemData.verificationQuestions = questions;
      }

      const result = await itemService.addItem(itemData);

      if (result.success) {
        alert('Your report has been successfully published!');
        onClose();
      } else {
        setError(result.error || 'Failed to submit report. Please try again.');
      }
    } catch (err) {
      console.error(err);
      setError('An error occurred during submission.');
    } finally {
      setPublishLoading(false);
    }
  };

  const handleQuestionChange = (index, field, value) => {
    const updated = [...questions];
    updated[index][field] = value;
    setQuestions(updated);
  };

  const renderStep1 = () => (
    <div>
      <h3 style={{ fontSize: '1.25rem', fontWeight: 600, marginBottom: '1rem' }}>Step 1: What type of report?</h3>
      
      <div style={{ display: 'flex', gap: '1rem', marginBottom: '2rem' }}>
        <button
          className="btn glass-interactive"
          style={{
            flex: 1,
            padding: '2rem 1rem',
            display: 'flex',
            flexDirection: 'column',
            gap: '0.5rem',
            border: type === 'found' ? '2px solid #00b894' : '1px solid var(--card-border)',
            background: type === 'found' ? 'rgba(0, 184, 148, 0.05)' : ''
          }}
          onClick={() => { setType('found'); handleRemoveImage(); }}
        >
          <span style={{ fontSize: '1.2rem', fontWeight: 700, color: '#00b894' }}>I Found an Item</span>
          <span style={{ fontSize: '0.8rem', color: 'var(--light-text)' }}>I want to verify the owner and return it</span>
        </button>

        <button
          className="btn glass-interactive"
          style={{
            flex: 1,
            padding: '2rem 1rem',
            display: 'flex',
            flexDirection: 'column',
            gap: '0.5rem',
            border: type === 'lost' ? '2px solid var(--accent-color)' : '1px solid var(--card-border)',
            background: type === 'lost' ? 'rgba(255, 118, 117, 0.05)' : ''
          }}
          onClick={() => { setType('lost'); handleRemoveImage(); }}
        >
          <span style={{ fontSize: '1.2rem', fontWeight: 700, color: 'var(--accent-color)' }}>I Lost an Item</span>
          <span style={{ fontSize: '0.8rem', color: 'var(--light-text)' }}>I want to search and matching list</span>
        </button>
      </div>

      <h3 style={{ fontSize: '1.25rem', fontWeight: 600, marginBottom: '1rem' }}>
        Upload Image {type === 'lost' && <span style={{ fontSize: '0.9rem', color: 'var(--light-text)', fontWeight: 400 }}>(Optional)</span>}
      </h3>

      {!imagePreview ? (
        <div className="upload-dropzone" onClick={() => fileInputRef.current.click()}>
          <Upload size={36} style={{ color: 'var(--primary-color)' }} />
          <div>
            <p style={{ fontWeight: 600 }}>Click to upload an image</p>
            <p style={{ fontSize: '0.8rem', color: 'var(--light-text)' }}>PNG, JPG or WEBP up to 5MB</p>
          </div>
          <input
            type="file"
            ref={fileInputRef}
            className="avatar-file-input"
            accept="image/*"
            onChange={handleImageChange}
          />
        </div>
      ) : (
        <div>
          <div className="preview-container">
            <img src={imagePreview} alt="Upload Preview" className="preview-image" />
            <button className="remove-image-btn" onClick={handleRemoveImage}>
              <X size={18} />
            </button>
          </div>

          {/* AI Image verification log */}
          {type === 'found' && moderationLoading && (
            <div className="ai-status-card ai-status-pending">
              <Loader2 className="spinner" size={16} style={{ width: 16, height: 16 }} />
              <span>Gemini AI is analyzing image rules (safe, no human faces)...</span>
            </div>
          )}

          {type === 'found' && moderationResult && moderationResult.verified && (
            <div className="ai-status-card ai-status-success">
              <ShieldCheck size={16} />
              <span>Gemini Verification: Image is safe and ready. AI questions generated!</span>
            </div>
          )}

          {type === 'found' && moderationResult && !moderationResult.verified && (
            <div className="ai-status-card ai-status-error">
              <AlertCircle size={16} />
              <span>Rejected: {moderationResult.reason || 'Image does not match rules.'}</span>
            </div>
          )}
        </div>
      )}
    </div>
  );

  const renderStep2 = () => (
    <div>
      <h3 style={{ fontSize: '1.25rem', fontWeight: 600, marginBottom: '1.5rem' }}>Step 2: Tell us about the item</h3>

      <div className="form-group">
        <label className="form-label">Item Title</label>
        <input
          type="text"
          className="form-input"
          placeholder="e.g. Black iPhone 13 Pro Max"
          value={title}
          onChange={(e) => { setTitle(e.target.value); setError(''); }}
          required
        />
      </div>

      <div className="form-group">
        <label className="form-label">Category</label>
        <select
          className="form-select"
          value={category}
          onChange={(e) => setCategory(e.target.value)}
        >
          {categories.map(cat => (
            <option key={cat} value={cat}>{cat}</option>
          ))}
        </select>
      </div>

      <div className="form-group">
        <label className="form-label">Description</label>
        <textarea
          className="form-textarea"
          rows="4"
          placeholder="Describe key details. (Keep specifics secret if it is a found item so that the verification questions remain useful!)"
          value={description}
          onChange={(e) => { setDescription(e.target.value); setError(''); }}
          required
        />
      </div>
    </div>
  );

  const renderStep3 = () => (
    <div>
      <h3 style={{ fontSize: '1.25rem', fontWeight: 600, marginBottom: '1.5rem' }}>Step 3: Where and how to reach?</h3>

      <div className="form-group">
        <label className="form-label">Location / Address</label>
        <input
          type="text"
          className="form-input"
          placeholder="e.g. Near CCD Cafe, Terminal 2 Airport"
          value={location}
          onChange={(e) => { setLocation(e.target.value); setError(''); }}
          required
        />
        <button
          type="button"
          className="btn btn-secondary"
          style={{ alignSelf: 'flex-start', padding: '0.4rem 0.8rem', fontSize: '0.8rem', marginTop: '0.5rem' }}
          onClick={() => {
            navigator.geolocation.getCurrentPosition(
              (pos) => {
                setLocation(`Lat: ${pos.coords.latitude.toFixed(4)}, Lon: ${pos.coords.longitude.toFixed(4)}`);
              },
              (err) => alert('Could not access current location.')
            );
          }}
        >
          Detect My Location
        </button>
      </div>

      <div className="form-group">
        <label className="form-label">Contact Mobile Number</label>
        <div style={{ display: 'flex', gap: '0.5rem' }}>
          <div className="form-input" style={{ width: '80px', display: 'flex', alignItems: 'center', justifyContent: 'center', background: '#eaeaea', fontWeight: 'bold' }}>
            +91
          </div>
          <input
            type="tel"
            className="form-input"
            placeholder="00000 00000"
            value={phone}
            onChange={(e) => { setPhone(e.target.value); setError(''); }}
            maxLength={10}
            required
          />
        </div>
      </div>
    </div>
  );

  const renderStep4 = () => (
    <div>
      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginBottom: '1rem' }}>
        <h3 style={{ fontSize: '1.25rem', fontWeight: 600 }}>Step 4: Owner Verification Questions</h3>
        {questionsLoading && (
          <span style={{ display: 'flex', alignItems: 'center', gap: '0.3rem', fontSize: '0.8rem', color: 'var(--primary-color)' }}>
            <Loader2 className="spinner" size={12} style={{ width: 12, height: 12 }} /> Generating...
          </span>
        )}
      </div>
      
      <p style={{ color: 'var(--light-text)', fontSize: '0.9rem', marginBottom: '1.5rem' }}>
        To prevent fraudulent claims, Gemini AI has auto-generated 3 specific questions based on the image you uploaded. You can customize them below.
      </p>

      {questions.map((q, idx) => (
        <div key={idx} className="question-item">
          <span style={{ fontWeight: 'bold', fontSize: '0.85rem', color: 'var(--primary-color)' }}>Question {idx + 1}</span>
          <input
            type="text"
            className="form-input"
            style={{ fontWeight: 500 }}
            placeholder={`Question ${idx + 1}`}
            value={q.q}
            onChange={(e) => handleQuestionChange(idx, 'q', e.target.value)}
          />
          <input
            type="text"
            className="form-input"
            placeholder={`Expected Answer for Question ${idx + 1}`}
            value={q.a}
            onChange={(e) => handleQuestionChange(idx, 'a', e.target.value)}
          />
        </div>
      ))}
    </div>
  );

  const renderStep5 = () => (
    <div style={{ textAlign: 'left' }}>
      <h3 style={{ fontSize: '1.25rem', fontWeight: 600, marginBottom: '1.5rem' }}>Step 5: Review & Publish</h3>
      
      <div className="glass" style={{ padding: '1.5rem', display: 'flex', flexDirection: 'column', gap: '1rem' }}>
        {imagePreview && (
          <img src={imagePreview} alt="Preview" style={{ width: '100%', height: '180px', objectFit: 'cover', borderRadius: 'var(--radius-sm)' }} />
        )}
        
        <div>
          <span className={`item-badge ${type === 'found' ? 'badge-found' : 'badge-lost'}`}>{type}</span>
          <h4 style={{ fontSize: '1.3rem', fontWeight: 700, margin: '0.5rem 0 0.2rem' }}>{title}</h4>
          <p style={{ fontSize: '0.85rem', color: 'var(--light-text)', fontWeight: 600 }}>Category: {category}</p>
        </div>

        <p style={{ color: 'var(--dark-text)', fontSize: '0.95rem', borderTop: '1px solid rgba(0,0,0,0.05)', paddingTop: '0.75rem' }}>
          {description}
        </p>

        <div style={{ borderTop: '1px solid rgba(0,0,0,0.05)', paddingTop: '0.75rem', display: 'flex', flexDirection: 'column', gap: '0.3rem', fontSize: '0.85rem', color: 'var(--light-text)' }}>
          <div><strong>Location:</strong> {location}</div>
          <div><strong>Contact:</strong> +91 {phone}</div>
        </div>

        {type === 'found' && (
          <div style={{ borderTop: '1px solid rgba(0,0,0,0.05)', paddingTop: '0.75rem' }}>
            <span style={{ fontWeight: 700, fontSize: '0.9rem', display: 'block', marginBottom: '0.5rem' }}>Verification Questions:</span>
            {questions.map((q, idx) => (
              <div key={idx} style={{ fontSize: '0.85rem', marginBottom: '0.25rem' }}>
                <strong>Q:</strong> {q.q}
              </div>
            ))}
          </div>
        )}
      </div>
    </div>
  );

  return (
    <div className="wizard-overlay">
      <div className="wizard-modal glass">
        
        <div className="wizard-header">
          <div style={{ display: 'flex', alignItems: 'center', gap: '0.5rem' }}>
            <Sparkles size={20} style={{ color: 'var(--primary-color)' }} />
            <h2 style={{ fontSize: '1.25rem', fontWeight: 700 }}>Report Lost/Found</h2>
          </div>
          <button style={{ background: 'none', border: 'none', cursor: 'pointer', color: 'var(--light-text)' }} onClick={onClose}>
            <X size={20} />
          </button>
        </div>

        <div className="wizard-steps-indicator">
          <div className={`step-indicator ${step === 1 ? 'active' : ''}`}>
            <span className="step-number">1</span>
            <span>Type</span>
          </div>
          <div className={`step-indicator ${step === 2 ? 'active' : ''}`}>
            <span className="step-number">2</span>
            <span>Details</span>
          </div>
          <div className={`step-indicator ${step === 3 ? 'active' : ''}`}>
            <span className="step-number">3</span>
            <span>Location</span>
          </div>
          {type === 'found' && (
            <div className={`step-indicator ${step === 4 ? 'active' : ''}`}>
              <span className="step-number">4</span>
              <span>Questions</span>
            </div>
          )}
          <div className={`step-indicator ${step === 5 ? 'active' : ''}`}>
            <span className="step-number">{type === 'found' ? 5 : 4}</span>
            <span>Submit</span>
          </div>
        </div>

        <div className="wizard-body">
          {error && (
            <div className="auth-error" style={{ marginBottom: '1.5rem' }}>
              <AlertCircle size={16} />
              <span>{error}</span>
            </div>
          )}

          {step === 1 && renderStep1()}
          {step === 2 && renderStep2()}
          {step === 3 && renderStep3()}
          {step === 4 && type === 'found' && renderStep4()}
          {step === 5 && renderStep5()}
        </div>

        <div className="wizard-footer">
          {step > 1 ? (
            <button className="btn btn-secondary" onClick={handleBack} disabled={publishLoading}>
              Back
            </button>
          ) : (
            <div />
          )}

          {step < 5 ? (
            <button 
              className="btn btn-primary" 
              onClick={handleNext} 
              disabled={type === 'found' && step === 1 && moderationLoading}
            >
              <span>Next Step</span>
              <ArrowRight size={16} />
            </button>
          ) : (
            <button className="btn btn-primary" onClick={handlePublish} disabled={publishLoading} style={{ background: '#00b894' }}>
              {publishLoading ? (
                <Loader2 className="spinner" size={16} style={{ width: 16, height: 16 }} />
              ) : (
                <span style={{ display: 'flex', alignItems: 'center', gap: '0.4rem' }}>
                  <Check size={16} /> Publish Report
                </span>
              )}
            </button>
          )}
        </div>

      </div>
    </div>
  );
}
